### ISSUE TITLE
Remove the TypeScript and Flow plugins, and add a --language typescript option that enables Babylon’s TypeScript option and disables its Flow option.

### ISSUE THREAD

**Issue by lipis** [prettier/prettier#4103]
From https://github.com/prettier/prettier/issues/3503#issuecomment-359863688

**Comment by j-f1** [comment]
We could also create `@prettier/plugin-typescript-eslint-parser` and `@prettier/plugin-flow-parser` packages to give people an alternative if they encounter bugs in Babylon’s implementation.

**Comment by JamesHenry** [comment]
Please keep me in the loop here, I am a prettier collaborator (I introduced TS support) and the maintainer of typescript-eslint-parser.

I've worked hard with the TypeScript team to align on ASTs, but there is still a little work left to be done. If we can keep communicating about how things are moving on each side then we can ensure the minimum possible duplication of work, and pain for users.

**Comment by suchipi** [comment]
FYI, it looks like some of TypeScript's new/upcoming features (like conditional types) can be parsed by the typescript parser but not by babylon.

**Comment by existentialism** [comment]
FWIW, conditional types landed in [babylon@7.0-beta.41](https://github.com/babel/babel/pull/7404) thanks to the TS team :)

**Comment by suchipi** [comment]
With that in mind, I'm a bit hesitant to switch to babylon for typescript in general... Not sure what the value-add of removing it is here. If we were starting from scratch, it'd be less AST nodes to care about, but we already have TypeScript working.

**Comment by j-f1** [comment]
@suchipi The value-add is removing a dependency, since we no longer have to bundle TypeScript’s parser, but we already bundle Babylon.

**Comment by duailibe** [comment]
@j-f1 I don't see what's the value in that

**Comment by j-f1** [comment]
@duailibe Prettier should get smaller and download faster.

**Comment by suchipi** [comment]
I don't think that is a valuable trade-off compared to TypeScript features getting to users sooner

**Comment by j-f1** [comment]
@suchipi How much of a delay is it really though, since we only release every month and a half?

**Comment by suchipi** [comment]
Good point, but I wouldn't want to tell users "we're waiting for typescript-eslint-parser/babel before we can add support for this" when we don't have to right now when using the typescript parser. The typescript parser itself is only loaded if you use it, and I don't think there's much value in making the tarball on npm smaller, especially considering it's only [1.8MB](https://arve0.github.io/npm-download-size/#prettier).

**Comment by j-f1** [comment]
@suchipi We do have to wait for TypeScript-ESLint-parser to add support right now, though:

https://github.com/prettier/prettier/blob/5776c8405a20f8f3026fa33c16fbd7676be8a19e/src/language-js/parser-typescript.js#L34-L45

**Comment by JamesHenry** [comment]
typescript-eslint-parser leverages the TypeScript compiler, so most things are automatically supported (have their ASTs converted).

**Comment by thorn0** [comment]
Compiling TS with Babel became much more popular since the last comment in this issue. E.g. Create React App does this. "[TypeScript features getting to users sooner](https://github.com/prettier/prettier/issues/4103#issuecomment-386815384)" is not a strong argument anymore as the TS team cooperates with the Babel team to make sure the parser is kept up to date. For example, here is the [issue](https://github.com/babel/babel/issues/10527) for the upcoming "assertion signatures" feature. On the other hand, we got stuck with updating `typescript-estree` as it doesn't support Node 6.

**Comment by fisker** [comment]
Postpone To v4.
